### PR TITLE
add evergreen integration

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
As discussed, a minimal working evergreen setup suffices for now, as long as it conducts static analysis and unit testing

### CODE REVIEW

**Review by jyemin — APPROVED** [review]
I assume you weren't able to test this, since there is no Evergreen project.  If so, let's just commit this and then iterate, once we see what the actually build looks like.

File: .evergreen/.evg.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+functions:
```
Comment (jyemin) [inline .evergreen/.evg.yml:1]: Rename this file to config.yml.  Java driver file name is not what most newer projects use.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: .evergreen/java-config.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+# Java configurations for evergreen
+
+export JDK8="/opt/java/jdk8"
```
Comment (jyemin) [inline .evergreen/java-config.sh:5]: Let's remove JDK8 and JDK11! :)

**Review by jyemin — CHANGES_REQUESTED** [review]

**Comment by NathanQingyangXu** [comment]
> I assume you weren't able to test this, since there is no Evergreen project. If so, let's just commit this and then iterate, once we see what the actually build looks like.

I could test locally by choosing the branch name (together with evergreen config file path) as below in the project settings page:

Screenshot 2024-10-18 at 10 21 15 AM

That is how I tested it previously. will try to investigate whether we could see the CI effect here by github integration option in the project settings as well.

**Comment by NathanQingyangXu** [comment]
I tested this branch indirectly by creating a separate temporary PR at https://github.com/mongodb/mongo-hibernate/pull/5. Now testing succeeded and I closed that testing PR. Latest commits in this PR reflected the issues flushed out during the testing.

File: .evergreen/config.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    - command: git.apply_patch
+
+  exec script:
+    - command: shell.exec
+      type: test
+      params:
+        working_dir: src
+        shell: bash
+        script: |
+          ${PREPARE_SHELL}
```
Comment (jyemin) [inline .evergreen/config.yml:17]: Doesn't this needs to be defined somewhere? I'm surprised this doesn't cause the script to error.

File: .evergreen/run-test.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+set -o xtrace   # Write all commands first to stderr
+set -o errexit  # Exit the script with error if any of the commands fail
+
+############################################
+#            Main Program                  #
+############################################
+source ./.evergreen/java-config.sh
+
+echo "mongo-hibernate: static checking ..."
```
Comment (jyemin) [inline .evergreen/run-test.sh:11]:
```suggestion
echo "mongo-hibernate: running tests ..."
```

File: .evergreen/config.yml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    commands:
+      - func: exec script
+        vars:
+          file: .evergreen/run-test.sh
+
+buildvariants:
+
+  - name: simple-variant
+    tags: ["ci"]
+    display_name: Simple Variant
```
Comment (jyemin) [inline .evergreen/config.yml:44]: Do we want all of these to have the same display name?  In Java driver, for example, the display name for static checks is "Static Checks".

**Review by jyemin — CHANGES_REQUESTED** [review]

File: .evergreen/config.yml (hunk 5 above)
Comment (NathanQingyangXu) [reply .evergreen/config.yml:44]: This display_name is for variant, not for task, so its counterpart is like: 

**4.2 Sharded Cluster NoAuth NoSSL JDK8 Linux**

File: .evergreen/config.yml (hunk 3 above)
Comment (NathanQingyangXu) [reply .evergreen/config.yml]: I removed it for currently it is not defined.
Guess shell might interpret unknown variable as empty string. For instance, the following command:
```
# echo $UNKNOWN_VARIABLE
```
simply output an empty string without complaining, unless `set -u` is run first.

File: .evergreen/config.yml (hunk 5 above)
Comment (NathanQingyangXu) [reply .evergreen/config.yml]: I changed it to concrete variant name instead. We'd better avoid ambiguous 'simple-variant' expression.

File: .evergreen/config.yml [hunk 6]
```diff
@@ -39,9 +38,9 @@ tasks:
 
 buildvariants:
 
-  - name: simple-variant
+  - name: linux-rhel
     tags: ["ci"]
-    display_name: Simple Variant
+    display_name: Redhat Linux
```
Comment (jyemin) [inline .evergreen/config.yml:43]: I'd still prefer just Linux until we feel the need to test on multiple linux variants (which we never will !).

**Review by jyemin — CHANGES_REQUESTED** [review]

**Review by jyemin — DISMISSED** [review]

**Comment by NathanQingyangXu** [comment]
I rebased on the main branch and made the following changes:

1. change palantir change to `palantirJavaFormat().formatJavadoc(true)`
2. delete the '.editorconfig' file for it was based on google-java-format. There is no .editorconfig file for palantir, so no bother

**Comment by jyemin** [comment]
FYI, if you force push like this, reviewers are not able to see what has changed since the last review.

File: gradle/libs.versions.toml [hunk 7]
```diff
@@ -4,7 +4,6 @@
 [versions]
 junit-jupiter = "5.10.3"
 spotless = "6.25.0"
-palantir = "2.50.0"
```
Comment (jyemin) [inline gradle/libs.versions.toml:7]: Why don't we need this?  Is it because it's picked up from Spotless?

File: build.gradle.kts [hunk 8]
```diff
@@ -62,7 +62,7 @@ spotless {
 
         removeUnusedImports()
 
-        palantirJavaFormat(libs.versions.palantir.get()).style("GOOGLE").formatJavadoc(true)
+        palantirJavaFormat().formatJavadoc(true)
```
Comment (jyemin) [inline build.gradle.kts:65]: As this doesn't have anything to do with Evergreen, can we do a separate PR for it?

**Review by jyemin — CHANGES_REQUESTED** [review]

**Comment by NathanQingyangXu** [comment]
yeah, a separate PR is justified. Will create one.

File: build.gradle.kts (hunk 8 above)
Comment (NathanQingyangXu) [reply build.gradle.kts:65]: yeah, will do.

**Review by jyemin — APPROVED** [review]
